src/store/indexeddb-local-backend.ts

```diff
@@ -401,11 +403,22 @@ export class LocalIndexedDBStoreBackend implements IIndexedDBBackend {
     public async syncToDatabase(userTuples: UserTuple[]): Promise<void> {
         const syncData = this.syncAccumulator.getJSON(true);
 
+        if (this.isPersisting) {
+            logger.warn("Skipping syncToDatabase() as persist already in flight");
+            this.pendingUserPresenceData.push(...userTuples);
+            return;
+        } else {
+            userTuples.unshift(...this.pendingUserPresenceData);
+            this.isPersisting = true;
+        }
+
         await Promise.all([
             this.persistUserPresenceEvents(userTuples),
             this.persistAccountData(syncData.accountData),
             this.persistSyncData(syncData.nextBatch, syncData.roomsData),
-        ]);
+        ]).finally(() => {
+            this.isPersisting = false;
+        });
```

Comment:
given the async-await we could try `try-finally`

Reply:
hm, i might be missing how finally interacts with async-await. how would try-finally help here?

Reply:
you're saying this could be `try { await Promise.all(); } finally { this.isPersisting = false }` - is this just stylistic? feels a bit weird to be mixing promise-chaining style with block-style just because there's an `await` in there? or is the intention to avoid ambiguity over the precedence of the `await`?

Reply:
We're tending towards async-await vs promise-chaining and I'm not seeing any promise chaining in this block. The ambiguity aspect is minimal. I will bring it up on the code style doc to codify something, for now this is merely a comment

Reply:
right, so i guess the concern is mixing imperative await-style with chain-style (.finally()). will switch to blocks for the `finally`